modules/xfeatures2d/src/daisy.cpp

```diff
@@ -0,0 +1,2057 @@
+    void gradient(T* im, int h, int w, T* dy, T* dx)
+    {
+      CV_Assert( dx != NULL );
+      CV_Assert( dy != NULL );
+
+      for( int y=0; y<h; y++ )
+      {
+         int yw = y*w;
+         for( int x=0; x<w; x++ )
+         {
+            int ind = yw+x;
+            // dx
+            if( x>0 && x<w-1 ) dx[ind] = (T) (((T)im[ind+1]-(T)im[ind-1])/2.0f);
+            if( x==0         ) dx[ind] = ((T)im[ind+1]-(T)im[ind]);
+            if( x==w-1       ) dx[ind] = ((T)im[ind  ]-(T)im[ind-1]);
+
+            //dy
+            if( y>0 && y<h-1 ) dy[ind] = (T) (((T)im[ind+w]-(T)im[ind-w])/2.0f);
+            if( y==0         ) dy[ind] = ((T)im[ind+w]-(T)im[ind]);
+            if( y==h-1       ) dy[ind] = ((T)im[ind]  -(T)im[ind-w]);
+         }
+      }
+    }
+
+    // be careful, 'data' is destroyed afterwards
+    template<class T> inline
+    //  original T* workspace=0 was removed
+    void layered_gradient( T* data, int h, int w, int layer_no, T* layers, int lwork = 0 )
+    {
+      int data_size = h * w;
+      CV_Assert(layers!=NULL);
+      memset(layers,0,sizeof(T)*data_size*layer_no);
+
+      bool was_empty = false;
+      T* work=NULL;
+      if( lwork < 3*data_size ) {
+         work = new T[3*data_size];
+         was_empty = true;
+      }
+
+      // // smooth the data matrix
+      // T* bdata = blur_gaussian_2d<T,T>( data, h, w, 0.5, 5, false);
+      float kernel[5]; gaussian_1d(kernel, 5, 0.5, 0);
+      memcpy( work, data, sizeof(T)*data_size );
+      convolve_sym( work, h, w, kernel, 5 );
+
+      T *dx = work+data_size;
+      T *dy = work+2*data_size;
+      gradient( work, h, w, dy, dx );
+
+#if defined _OPENMP
+#pragma omp parallel for
+#endif
+      for( int l=0; l<layer_no; l++ )
+      {
+         float angle = (float) (2*l*CV_PI/layer_no);
+         float kos = (float) cos( angle );
+         float zin = (float) sin( angle );
+
+         T* layer_l = layers + l*data_size;
+
+         for( int index=0; index<data_size; index++ )
+         {
+            float value = kos * dx[ index ] + zin * dy[ index ];
+            if( value > 0 ) layer_l[index] = value;
+            else            layer_l[index] = 0;
+         }
+      }
+      if( was_empty ) delete []work;
+    }
+
+    // casts a type T2 array into a type T1 array.
+    template<class T1, class T2> inline
+    T1* type_cast(T2* data, int sz)
+    {
+      T1* out = new T1[sz];
+
+      for( int i=0; i<sz; i++ )
+         out[i] = (T1)data[i];
+
+      return out;
+    }
+
+    // Applies a 2d gaussian blur of sigma std to the input array.  if
+    // kernel_size is not set or it is set to 0, then it is taken as
+    // 3*sigma and if it is set to an even number, it is incremented
+    // to be an odd number.  if in_place=true, then T1 must be equal
+    // to T2 naturally.
+    template<class T1, class T2> inline
+    T1* blur_gaussian_2d( T2* array, int rn, int cn, float sigma, int kernel_size = 0, bool in_place = false )
```

Comment:
[DONE] but with 2 x filter2D to be 1:1 bitwise level with same results as original.